cmd/kops-controller/pkg/metrics/collector/validate.go

```diff
@@ -0,0 +1,240 @@
+	unhealthyComponents := []v1.ComponentStatus{}
+
+	// TODO: ComponentStatuses API has been deprecated since 1.19, so we have to rewrite this API call.
+	// https://github.com/kubernetes/kops/issues/10057
+	componentList, err := c.k8sClient.CoreV1().ComponentStatuses().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return healthyComponents, unhealthyComponents, err
+	}
+
+	for _, component := range componentList.Items {
+		con := true
+		for _, condition := range component.Conditions {
+			if condition.Status != v1.ConditionTrue {
+				con = false
+			}
+		}
+		if con {
+			healthyComponents = append(healthyComponents, component)
+		} else {
+			unhealthyComponents = append(unhealthyComponents, component)
+		}
+	}
+	return healthyComponents, unhealthyComponents, nil
+}
+
+var masterStaticPods = []string{
+	"kube-apiserver",
+	"kube-controller-manager",
+	"kube-scheduler",
+}
+
+func (c *validateCollector) pods(ctx context.Context, nodes []v1.Node) ([]v1.Pod, []v1.Pod, []v1.Pod, map[string]string, error) {
+	masterWithoutPod := map[string]map[string]bool{}
+	nodeByAddress := map[string]string{}
+	for _, node := range nodes {
+		labels := node.GetLabels()
+		if labels != nil && labels["kubernetes.io/role"] == "master" {
+			masterWithoutPod[node.Name] = map[string]bool{}
+			for _, pod := range masterStaticPods {
+				masterWithoutPod[node.Name][pod] = true
+			}
+		}
+		for _, nodeAddress := range node.Status.Addresses {
+			nodeByAddress[nodeAddress.Address] = node.Name
+		}
+	}
+
+	pendingPods := []v1.Pod{}
+	unknownPods := []v1.Pod{}
+	notReadyPods := []v1.Pod{}
+	err := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
+		return c.k8sClient.CoreV1().Pods(metav1.NamespaceAll).List(ctx, opts)
+	})).EachListItem(context.TODO(), metav1.ListOptions{}, func(obj runtime.Object) error {
+		pod := obj.(*v1.Pod)
+
+		app := pod.GetLabels()["k8s-app"]
+		if pod.Namespace == "kube-system" && masterWithoutPod[nodeByAddress[pod.Status.HostIP]][app] {
+			delete(masterWithoutPod[nodeByAddress[pod.Status.HostIP]], app)
+		}
+
+		priority := pod.Spec.PriorityClassName
+		if priority != "system-cluster-critical" && priority != "system-node-critical" {
+			return nil
+		}
+		if pod.Status.Phase == v1.PodSucceeded {
+			return nil
+		}
+		if pod.Status.Phase == v1.PodPending {
+			pendingPods = append(pendingPods, *pod)
+			return nil
+		}
+		if pod.Status.Phase == v1.PodUnknown {
+			unknownPods = append(unknownPods, *pod)
+			return nil
+		}
+		for _, container := range pod.Status.ContainerStatuses {
+			if !container.Ready {
+				notReadyPods = append(notReadyPods, *pod)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("error listing Pods: %v", err)
+	}
+
+	missingPod := map[string]string{}
+	for node, nodeMap := range masterWithoutPod {
+		for app := range nodeMap {
+			missingPod[app] = node
+		}
+	}
+
+	return pendingPods, unknownPods, notReadyPods, missingPod, nil
+}
```

Comment:
These codes are almost same as https://github.com/h3poteto/kops/blob/master/pkg/validation/validate_cluster.go .
I think that I should use same functions as `validate_cluster,go` , but for that I need to refactor `validate_cluster.go` . If I contain that changes in this pull request, changes too big. So I will refactor this codes after this pull request is merged.

Reply:
I'm not a fan of duplicating code like this. If you need to refactor, then do that first. But why can't you just call validate, then classify the returned failures?

Reply:
I can't refactor it before implement prometheus exporter, because I don't know which methods in `validate_cluster` is required in prometheus exporter. If you don't care about the amount of change, I can refactor it now.

Validator method only returns `ValidationError` struct: https://github.com/kubernetes/kops/blob/55c5c5cf0987a714e2dc7db9fc19b284029c4d61/pkg/validation/validate_cluster.go#L47-L53
It contains only validation error message, so I can't classify it by the required metrics. This struct is designed with the intention of producing errors on stdout, it is not match prometheus metrics.